### ISSUE TITLE
Flutter frames not rendering propperly

### ISSUE THREAD

**Issue by zigapk** [wmww/gtk-layer-shell#71]
Hi there!

This might not be the right place for this question, but I couldn't come up with a better one.

I'm trying to modify Flutter for Linux GTK code to run my app and anchor it inside sway shell using layer-shell protocol.

It seems to work fine at the first glance (see image below). However, no frames are rendered after the first one. I've spent quite some time trying to figure out why this would be the case but have no idea.

I'd be glad if you could come up with any ideas, otherwise feel free to close the issue. Thanks a lot!

![image](https://user-images.githubusercontent.com/5074514/96728810-bfedb600-13b4-11eb-93d8-5584860f352e.png)

**Comment by wmww** [comment]
Likely due to https://github.com/swaywm/sway/issues/5617. On Wayland, Flutter creates and manages subsurface for the Flutter content of the window. Should work on non-wlroots compositors that support Layer Shell (of which Mir is the only one I know). Try it on mir_demo_server (`sudo apt install mir-test-tools` on Ubuntu) or the Mir-based mate-wayland session prototype (`sudo snap install mate-wayland --edge --classic`). If it works there, it's the Sway bug.

**Comment by zigapk** [comment]
Thank you for your suggestion.

Unfortunately I couldn't get any Flutter app to run Mir using Wayland backend - not even Flutter Gallery. I'm on Arch using self compiled Mir, so there is some room for improvement. I'll try official package on Ubuntu later today.

I've also pushed an [example](https://github.com/zigapk/flutter-layer-shell-demo/blob/master/linux/my_application.cc#L18-L38), illustrating my issues. This is an additional explanation of my issue, rather than a reply to your comment.

I do realize this is no responsibility of yours and don't expect you to spend any significant time on this. However, being involved in Flutter engine, Mir and Sway gives you a unique perspective on this and I am grateful for your help.

**Comment by ammen99** [comment]
@zigapk If you already have Sway working, then most probably you will have little difficulty in getting Wayfire to work. Wayfire does not suffer from https://github.com/swaywm/sway/issues/5617, so that way you will be able to check whether it is that bug or not.

**Comment by wmww** [comment]
> Unfortunately I couldn't get any Flutter app to run Mir using Wayland backend

I'd like to dig into what went wrong in case there's an underlying Mir problem (either actual bug or bad UX). I don't have my Arch box with me at the moment, but I do regularly compile and run Mir on Arch without issue. Were you able to successfully build Mir? Were you able to run other Wayland apps in Mir (such as, say, gedit or another GTK3 app)? Try this and see if any errors are shown (`wayland-10` is arbitrarily chosen as a display unlikely to conflict with any other on the system):
```
WAYLAND_DISPLAY=wayland-10 ./mir/build/bin/miral-shell
```
Then, in another terminal outside of Mir while Mir is running
```
WAYLAND_DISPLAY=wayland-10 flutter_app
```
I'm interested what you were trying differently if that does work, and what the errors are if it doesn't.

**Comment by zigapk** [comment]
Turns out it took me some time to setup a physical Ubuntu machine, sorry for the delay. I've finally got some answers:

1. I reinstalled the `mir-git` AUR package on Arch and both `miral-shell` and `mir_demo_server` now work. No idea what changed, I'll look through my bash history and try to find out.
2. I've tried the Flutter app on `mate-wayland` on Ubuntu 20.04 and it works.
3. The app also works on `mir_demo_server` on Arch.
4. `miral-shell` does not seem, to support `layer-shell-protocol` which I assume is expected too.

To summarize: this issue does seem to be related to swaywm/sway#5617, as I've also noticed the surface flickering sometimes, as I move the other apps around the workspace.

I couldn't get `mir_demo_server` to start on Ubuntu 20.04 getting the same error as MirServer/mir#914. I definitely wasn't running it in a VM. This is not particularly relevant to this discussion, I just wanted to provide some further information regarding ther mir issue in case it helps.

**Comment by wmww** [comment]
Thanks.

You are correct that `miral-shell` doesn't support layer shell. Support can be added with `--add-wayland-extensions zwlr_layer_shell_v1`. This will only work on a recent build of master. You don't have to test this, I just want to let you know it exists.

> I couldn't get mir_demo_server to start on Ubuntu 20.04

Was this after installing it from apt, a snap or building from source? If installed, which mir-related package(s) did you install?

**Comment by zigapk** [comment]
I can start both `mate-wayland` and `mir_demo_server` nested within gnome on W`ayland and X. However I can't start either from the display manager (gdm and lightdm) or a separate tty. Am I missing something?

I've installed `mir-test-tools` via apt and `mir-wayland` via snap.

**Comment by wmww** [comment]
Hmm, if you don't mind, try `sudo apt install mir-platform-graphics-gbm-kms18` and then try `mir_demo_server` from a TTY again (if everything goes black and you're able to exit back to the TTY with Ctrl+Alt+Backspace, you know it worked). If that does fix the issue, I'm not sure why that's not installed by default.

**Comment by zigapk** [comment]
It turns out that I forgot I had two displays connected. None of those servers seems to manage that, but I suppose this is expected? Anyway, no issues on the mir side.
